### PR TITLE
Dts - Device Tree Output

### PR DESCRIPTION
**Pull request by terpstra**
This adds support to rocketchip for generating a compliant device tree description. This should resolve issues #516 #474 #232, once we update spike+bbl+linux.

### CODE REVIEW

**Comment by terpstra** [comment]
Remaining to be done: sort out the boot ROM contents. The ROM should set a0 to hartid and a1 to a pointer to the compiled and flattened dts. Right now the ROM behaves analogous to the old config string approach.

File: src/main/scala/diplomacy/Nodes.scala [hunk 1]
```diff
@@ -302,8 +302,8 @@ class MixedNexusNode[DI, UI, EI, BI <: Data, DO, UO, EO, BO <: Data](
   numPI: Range.Inclusive = 1 to 999)
   extends MixedNode(inner, outer)(numPO, numPI)
 {
-  require (numPO.end >= 1, s"${name} does not accept outputs${lazyModule.line}")
-  require (numPI.end >= 1, s"${name} does not accept inputs${lazyModule.line}")
+//  require (numPO.end >= 1, s"${name} does not accept outputs${lazyModule.line}")
+//  require (numPI.end >= 1, s"${name} does not accept inputs${lazyModule.line}")
```
Comment (aswaterman) [inline src/main/scala/diplomacy/Nodes.scala]: Remove commented code?

**Comment by neuschaefer** [comment]
Wouldn't DTB/FDT (devicetree binary/blob, aka. flattened devicetree) be a better format than DTS here? Operating systems already have DTB parsers, but AFAIK no DTS parsers. And if a user wants to look at the devicetree in DTS form, they can still easily decompile it from DTB or linux's `/proc/device-tree` using the [devicetree compiler](https://git.kernel.org/cgit/utils/dtc/dtc.git).

In your current plans, will the bootloader or the kernel have to understand DTS? IOW, will the bootloader compile the DTS to DTB so it's easier to parse for the kernel?

**Comment by terpstra** [comment]
This is an intermediate commit. I need to revise how the ROM is generated still. The end goal is DTB and a pointer in a1 on power-up.

**Comment by neuschaefer** [comment]
Ah, ok, I'm totally fine with that. Thanks.

**Pull request merged**